modules/secure-preview/token.php

```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ * The token module for securing previews.
+ *
+ * @package vip-workflow
+ */
+namespace VIPWorkflow\Modules\SecurePreview;
+
+use WP_Error;
+
+class Token {
+	const META_KEY     = 'vw_secure_preview_token';
+	const TOKEN_ACTION = 'preview';
+
+	/**
+	 * Generate a one-time-use authentication token that can be returned with a
+	 * subsequent request and validated. Upon verification, both the token and the
+	 * expiration should be validated. (See validate_token.)
+	 *
+	 * @param  int    $post_id            Post ID.
+	 * @param  string $cap                Capability required to generate this token.
+	 * @param  string $is_one_time_use    Whether this token should be deleted after one use.
+	 * @param  string $expiration_seconds The number of seconds this token should be valid for.
+	 * @return array|WP_Error
+	 */
+	public static function generate_token( $post_id, $cap, $is_one_time_use, $expiration_seconds ) {
+		if ( ! current_user_can( $cap, $post_id ) ) {
+			return new WP_Error( 'vip-workflow-token-permission-denied', __( 'You do not have sufficient permissions to access this page.', 'vip-workflow' ) );
+		}
+
+		$token_lifetime = self::get_token_lifetime_in_seconds( $expiration_seconds );
+		$expiration     = time() + $token_lifetime;
+		$secret         = wp_generate_password( 64, true, true );
+		$token          = hash_hmac( 'sha256', self::TOKEN_ACTION, $secret );
+
+		$meta_value = [
+			'action'       => self::TOKEN_ACTION,
+			'expiration'   => $expiration,
+			'one_time_use' => $is_one_time_use,
+			'user'         => get_current_user_id(), // not validated, just stored in case it's interesting.
+			'token'        => $token,
+			'version'      => 1,                     // not validated, but might be useful in future.
+		];
+
+		add_post_meta( $post_id, self::META_KEY, $meta_value );
+
+		return $token;
+	}
+
+	/**
+	 * Validate a token that has been sent with a request.
+	 *
+	 * @param  string $token   Token.
+	 * @param  int    $post_id Post ID.
+	 * @return bool
+	 */
+	public static function validate_token( $token, $post_id ) {
+		$meta_key     = self::META_KEY;
+		$saved_tokens = get_post_meta( $post_id, $meta_key );
```

Comment:
`false` is the default, but a good signal that we expect multiple
```suggestion
		$saved_tokens = get_post_meta( $post_id, $meta_key, false );
```

Reply:
Addressed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/4a920c7561c0ac05605fd7cf19c22ea9ad72227d. Curiously I also had to add a type check because explicitly adding `false` triggered [a sniff error](https://github.com/Automattic/VIP-Coding-Standards/blob/develop/WordPressVIPMinimum/Sniffs/Functions/CheckReturnValueSniff.php):

![Screenshot 2024-08-16 at 11 29 08 AM](https://github.com/user-attachments/assets/8f7ce3d5-e60c-4c7a-8da7-0a72780743a3)